plat: Migrate bus drivers to drivers/ukbus

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]

### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Migrate the PCI and platform drivers to `drivers/ukbus/`

GitHub-Depends-On: #1023

michpappas commented:
~~Notice that the driver is introduced as a platform library, ie:~~
```
$(eval $(call addplatlib_s,kvm,libukbus_pci,$(CONFIG_LIBUKBUS_PCI)))
```
~~This is due to limitations of #1023. **Do not merge until resolved.**~~

michpappas commented:
Changes in this update:
- Rebase to latest state of #1023
- Replace subsystem's `Config.uk` `menuconfig` with `menu`
- Register library using `addlib_s`
- Add `exportsyms.uk`
- Add `ASINCLUDES` and `CSSINCLUDES`

michpappas committed:
plat/kvm: Enable the platform bus by default on arm64

Enable the platform bus driver on all KVM VMMs for arm64.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers: Update includes to the new platform bus headers path

With the platform bus driver being migrated to the ukbus drivers subsystem
its headers path and names have been updated according the the drivers
conventions. Update drivers to use the new include path.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers: Update includes to the new PCI bus driver headers path

With the PCI driver being migrated to the ukbus drivers subsystem
its headers path and names have been updated according the the drivers
conventions. Update drivers to use the new include path.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm: Enable the PCI driver by default

Enable LIBUKBUS_PCI on all KVM VMMs. That will be eventually replaced
by VIRTIO_PCI, yet we add it now to maintain backwards compatibility
with existing applications.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat: Migrate the platform bus driver to drivers/ukbus/

Migrate the platform bus driver to the newly introduced ukbus drivers
subsystem.

Notice that the migration of this driver is performed in two steps:

1. Migrate the sources into driver verbatim (does not compile)
2. Update the sources for this migration (next commit)

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukbus/pci: Update include guards

Update include guards to the drivers conventions and add guards for
linkage with C++.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukbus/platform: Update include guards

Update include guards to the drivers convention and add guards for
linkage with C++.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukbus/pci: Fix result check of drv->add_dev() operation

Fix a bug in pci_ecam where a check against the result of drv->add_dev()
would happen outside the scope of the conditional call to that function.

This would trigger failure of arch_pci_driver_add_device() depending on
the state of the stack.

This bug was pinpointed by a long ignored GCC warning:

  warning: ‘ret’ may be used uninitialized in this function [-Wmaybe-uninitialized]

Checkpatch-Ignore: COMMIT_LOG_LONG_LINE
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukbus/pci: Memset dummy_imask using size not element count

The call of memset in the initialization of dummy_imask in the fdt
irq parser of pci_ecam would pass the number of array elements instead
of the array size in bytes.

This bug was pinpointed by a long ignored GCC warning:

  warning: ‘memset’ used with length equal to number of elements without multiplication by element size [-Wmemset-elt-size]
  237 |  memset((void *)dummy_imask, cpu_to_fdt32(~0), ARRAY_SIZE(dummy_imask));
      |  ^~~~~~

Checkpatch-Ignore: COMMIT_LOG_LONG_LINE
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented on drivers/ukbus/pci/arch/arm64/pci_bus.c:
```diff
@@ -95,14 +95,16 @@ static int arch_pci_driver_add_device(struct pci_driver *drv,
-			  (int) addr->devid,
-			  (int) addr->function);
-		uk_free(pha, dev);
+		if (ret < 0) {
```
Perhaps an `unlikely` here? :thinking:

mogasergiu commented on drivers/ukbus/pci/pci_ecam.c:
```diff
@@ -234,7 +234,7 @@ int gen_pci_irq_parse(const fdt32_t *addr, struct fdt_phandle_args *out_irq)
 
 	ipar = gen_pci_fdt;
-	memset((void *)dummy_imask, cpu_to_fdt32(~0), ARRAY_SIZE(dummy_imask));
+	memset((void *)dummy_imask, cpu_to_fdt32(~0), sizeof(dummy_imask));
```
commit message headline too long, it appears as
```
drivers/ukbus/pci: Fix initialization of dummy_imask in gen_pci_irq_p…
…arse
```

I see that you added
```
Checkpatch-Ignore: COMMIT_LOG_LONG_LINE
```
Maybe it would be worth re-wording it though as this is too generic and not specific enough. How about:
```
drivers/ukbus/pci: Memset `dummy_imask` by size not unit count
```
I've also been met with similar situations quite frequently unfortunately :disappointed:.

michpappas replied:
Actually the headline is 73 chars, while the checkpatch limit is 75. The problem is the error message of GCC. I will wrap the long line (~120 chars) so that the max line is 80 chars.

mogasergiu replied:
> Actually the headline is 73 chars, while the checkpatch limit is 75.

Which is still above 70, the maximum recommended for headlines AFAIK. Also, it is too generic, but I do not have any strong feelings towards this, keep it as you like.

> The problem is the error message of GCC. I will wrap the long line (~120 chars) so that the max line is 80 chars.

I would actually leave that as is. From what I have seen on mailing lists, usually errors/warnings/crashdumps/etc are kept without wrapping

michpappas replied:
This is the relevant snippet from checkpatch:
```
# Check for line lengths > 75 in commit log, warn once
                if ($in_commit_log && !$commit_log_long_line &&
                    length($line) > 75 &&
                    !($line =~ /^\s*[a-zA-Z0-9_\/\.]+\s+\|\s+\d+/ ||
                                        # file delta changes
                      $line =~ /^\s*(?:[\w\.\-]+\/)++[\w\.\-]+:/ ||
                                        # filename then :
                      $line =~ /^\s*(?:Fixes:|Link:)/i ||
                                        # A Fixes: or Link: line
                      $commit_log_possible_stack_dump)) {
                        WARN("COMMIT_LOG_LONG_LINE",
                             "Possible unwrapped commit description (prefer a maximum 75 chars per line)\n" . $herecurr);
                        $commit_log_long_line = 1;
                }
```

>  I would actually leave that as is. From what I have seen on mailing lists, usually errors/warnings/crashdumps/etc are kept without wrapping

Yes, that was my initial rationale. Ok then I'll leave it and we can blame GCC for spewing such long messages.

mogasergiu commented on plat/kvm/Config.uk:
```diff
@@ -10,7 +10,8 @@ menuconfig PLAT_KVM
        select LIBUKOFW if ARCH_ARM_64
-	   select LIBUKINTCTLR if ARCH_X86_64
+       imply LIBUKBUS_PCI if LIBUKBUS
+       select LIBUKINTCTLR if ARCH_X86_64
```
Why `if ARCH_X86_64`?

michpappas replied:
This is part of a different PR, I only fixed indentation here. But since you ask: Normally the platform would choose a specific driver (eg PIC, x2APIC, GICv2, GICv3), which in turn would ~~select~~ `depend on` the API library, ie LIBUKINTCTLR. However, right now on x86 we don't support this yet, so we enable the library that provides the API. And on Arm we need to select the GIC driver at the VMM level, as QEMU virt by default uses GICv2 while FC uses GICv3.

mogasergiu replied:
Ah, I see makes perfect sense, so this is temporary for x86 :+1:

mogasergiu reviewed (commented):
Had a quick run, will do a round of testing as well after my questions have been addressed and another for reviewing as well. Otherwise I greatly appreciate the fact that you did take my opinion into account regarding commit layout, thank you :pray:.

michpappas committed:
plat/kvm: Update Kconfig options to use LIBUKBUS_PCI

The migration of PCI to the drivers subsystem has caused the update
of its Kconfig options according to the driver's conventions. Update
plat/kvm/Config.uk accordingly.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

rares-miculescu commented on drivers/ukbus/pci/Makefile.uk:
```diff
@@ -0,0 +1,12 @@
+CXXINCLUDES-$(CONFIG_LIBUKBUS_PCI) += -I$(LIBUKBUS_PCI_BASE)/include
+
+LIBUKBUS_PCI-CINCLUDES-$(CONFIG_LIBUKBUS_PCI) += -I$(LIBUKBUS_PCI_BASE)/include
+
```
I think you should also include the platform headers, because `pci_ecam.c` has `#include <uk/bus/platform.h>`.

rares-miculescu reviewed (approved):
All good on my side. Thank you, @michpappas!

Reviewed-by: Rares Miculescu [miculescur@gmail.com](mailto:miculescur@gmail.com)

michpappas commented:
Pushed an update to address an issue identified by @rares-miculescu where enabling `CONFIG_LIBFDT` on x86 causes a compile error:

```
/home/mpp/devel/unikraft/uk_helloworld/unikraft/drivers/ukbus/pci/pci_ecam.c:34:10: fatal error: uk/bus/platform.h: No such file or directory
   34 | #include <uk/bus/platform.h>
      |          ^~~~~~~~~~~~~~~~~~~
```

The issue is caused by:
1. `libfdt` being visible on `x86_64`
2. `pci_ecam.c` incorrectly depending only on `CONFIG_LIBFDT`, given that it also depends on `CONFIG_LIBUKBUS_PLATFORM` (which is only enabled on Arm by the platform).

I pushed an update that introduces an invisible option `LIBUKBUS_PCI_ECAM` that is enabled conditionally to the above dependencies.

michpappas committed:
plat: Migrate PCI bus driver to drivers/ukbus/

Introduce the ukbus drivers subsystem and migrate the PCI driver from
plat/common.

Notice that the migration of this driver is performed in two steps:

1. Migrate the sources into driver verbatim (does not compile)
2. Update the sources for this migration (next commit)

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented on drivers/ukbus/pci/pci_ecam.c:
```diff
@@ -234,7 +234,7 @@ int gen_pci_irq_parse(const fdt32_t *addr, struct fdt_phandle_args *out_irq)
 
 	ipar = gen_pci_fdt;
-	memset((void *)dummy_imask, cpu_to_fdt32(~0), ARRAY_SIZE(dummy_imask));
+	memset((void *)dummy_imask, cpu_to_fdt32(~0), sizeof(dummy_imask));
```
For the commit message:
`s/initializaztion/initialization/`

mogasergiu commented on drivers/ukbus/pci/arch/arm64/pci_bus.c:
```diff
@@ -95,14 +95,16 @@ static int arch_pci_driver_add_device(struct pci_driver *drv,
-		uk_free(pha, dev);
+		if (unlikely(ret < 0)) {
+			uk_pr_err("PCI %02x:%02x.%02x: Failed to initialize device driver\n",
+				  (int) addr->bus,
```
I know this was present here before, but the only other `checkpatch.pl --strict` `CHECK` I am getting in this PR is regarding the whitespace after the cast. I know this is going to be deleted anyway when it will be reworked so probably no point in pointing this out now, but I thought I'd let you know in case that maybe you missed this and you would have wanted to modify it regardless. I leave it up to you if you want to waste some more time modifying this or not.

mogasergiu reviewed (approved):
This all looks good to me. Just two tiny comments. Nevertheless, this is great important work that I can't wait to get merged in its entirety. Make Unikraft great again, driver infrastructure coming in! :truck: 

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

michpappas committed:
drivers/ukbus/pci: Fix checkpatch styling issues

Fix some coding style issues reported by checkpatch
about space after cast on arm64.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
